rust/lance-index/src/vector/hnsw.rs

```diff
@@ -17,21 +17,146 @@
 //! Hierarchical Navigable Small World (HNSW).
 //!
 
+use std::collections::{BTreeMap, HashMap};
+use std::fmt::Debug;
+use std::ops::Range;
 use std::sync::Arc;
-use std::{collections::BTreeMap, fmt::Debug};
 
-use lance_core::Result;
+use arrow_array::{
+    builder::{ListBuilder, UInt32Builder},
+    cast::AsArray,
+    ListArray, RecordBatch, UInt32Array,
+};
+use arrow_schema::{DataType, Field, Schema, SchemaRef};
+use lance_core::{Error, Result};
+use lance_file::{reader::FileReader, writer::FileWriter};
 use lance_linalg::distance::MetricType;
+use lance_table::io::manifest::ManifestDescribing;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use snafu::{location, Location};
 
-use super::graph::{Graph, OrderedFloat};
+use super::graph::{
+    builder::GraphBuilder, storage::VectorStorage, Graph, OrderedFloat, NEIGHBORS_COL,
+    NEIGHBORS_FIELD,
+};
 use crate::vector::graph::beam_search;
 pub mod builder;
 pub use builder::HNSWBuilder;
 
+const HNSW_TYPE: &str = "HNSW";
+const POINTER_COL: &str = "__pointer";
+
+lazy_static::lazy_static! {
+    /// POINTER field.
+    ///
+    pub static ref POINTER_FIELD: Field = Field::new(POINTER_COL, DataType::UInt32, true);
+}
+
+/// One level of the HNSW graph.
+///
+/// Type parameters:
+/// ----------------
+/// T : Float
+///   The data type of the vectors?
```

Comment:
Why the `?`

Is that because `T` is unused at the moment?

Reply:
Yea i should remove this one.